File: server/php/src/Presentation/Controllers/PaymentLinkController.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Controllers;

use Fig\Http\Message\StatusCodeInterface;
use OnlinePayments\ExampleApp\Application\Interfaces\Services\PaymentLinkServiceInterface;
use OnlinePayments\ExampleApp\Configuration\Exceptions\ValidationException;
use OnlinePayments\ExampleApp\Presentation\Extensions\ResponseExtension;
use OnlinePayments\ExampleApp\Presentation\Mappers\PaymentLinkMapper;
use OnlinePayments\ExampleApp\Presentation\Models\PaymentLink\Request;
use OnlinePayments\ExampleApp\Presentation\Validators\PaymentLink\PaymentLinkValidator;
use Slim\Psr7\Request as SlimRequest;
use Slim\Psr7\Response as SlimResponse;

final readonly class PaymentLinkController
{
    public function __construct(
        private PaymentLinkServiceInterface $paymentLinkService,
        private PaymentLinkValidator        $paymentLinkValidator
    ) {}

    /**
     * @throws ValidationException
     */
    public function createPaymentLink(SlimRequest $request): SlimResponse
    {
        $presentationRequest = Request::fromRequest($request);

        $this->paymentLinkValidator->validate($presentationRequest);

        $responseDto = $this->paymentLinkService->createPaymentLink(PaymentLinkMapper::map($presentationRequest));

        $presentationResponse = PaymentLinkMapper::mapResponse($responseDto);

        return ResponseExtension::createSlimResponse($presentationResponse, StatusCodeInterface::STATUS_CREATED);
    }
}

File: server/python/app/presentation/controllers/payment_link_controller.py
from fastapi import APIRouter, Depends
from dependency_injector.wiring import inject, Provide
from fastapi import status
from app.application.interfaces.services.payment_link_service_interface import PaymentLinkServiceInterface
from app.configuration.di_container.di_container import Container
from app.presentation.mappers.payment_link_mapper import PaymentLinkPresentationMapper
from app.presentation.models.payment_link.request import PaymentLinkRequest
from app.presentation.models.payment_link.response import PaymentLinkResponse
from app.presentation.validators.payment_link.payment_link_validator import PaymentLinkValidator

router = APIRouter(tags=["Payment Link"])


@router.post("/links", response_model=PaymentLinkResponse, status_code=status.HTTP_201_CREATED)
@inject
def create_payment_link(
        request: PaymentLinkRequest,
        service: PaymentLinkServiceInterface = Depends(Provide[Container.payment_link_service])
):
    PaymentLinkValidator().validate(request)

    return PaymentLinkPresentationMapper.from_dto(
        service.create_payment_link(PaymentLinkPresentationMapper.to_dto(request)))
